### PR TITLE
Add workqueue example

### PR DESCRIPTION
**Pull request by rmohr**
Demonstrates how to compose a controller out of cache.Controller, cache.Indexer and a workqueue.

Tested on kubernetes 1.4.5.

### CODE REVIEW

**Comment by rmohr** [comment]
cc @caesarxuchao @wfarr

File: examples/workqueue/main.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	}, cache.Indexers{})
+
+	controller := NewController(queue, indexer, informer, func(cache.Indexer, workqueue.RateLimitingInterface) bool {
+		// Wait until there is a new item in the working queue
+		key, quit := queue.Get()
+		if quit {
+			return false
+		}
+		// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+		// This allows save parallel processing because two pods with the same key are never processed in
```
Comment (lavalamp) [inline examples/workqueue/main.go]: s/save/safe/

File: examples/workqueue/main.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+			// Tricky what to do in this situation. One thing we can do, is enqueueing it a few times to
+			// add some backoff delays on the invalid key. This way we avoid hotlooping
+			// on invalid keys. Furhter let's now and then forget about the key, so that the queue
+			// does not get too big in case the queue is flooded with a log of invalid keys.
+			if queue.NumRequeues(key) < 20 {
+				queue.AddRateLimited(key)
+			} else {
+				queue.Forget(key)
+			}
+			glog.Fatalf("Invalid key %s", key.(string))
```
Comment (lavalamp) [inline examples/workqueue/main.go]: this line seems wrong

File: examples/workqueue/main.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			return true
+		}
+
+		if !exists {
+			// Below we will warm up our cache with a Pod, so that we will see a delete for one pod
+			fmt.Printf("Pod %s does not exist anymore\n", key.(string))
```
Comment (lavalamp) [inline examples/workqueue/main.go]: Use glog for log messages (or the built-in log package).

File: examples/workqueue/main.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+	// Now let's start the controller
+	stop := make(chan struct{})
+	defer close(stop)
+	go controller.Run(1, stop)
+
+	// Wait forever
+	for {
+		time.Sleep(60000 * time.Millisecond)
+	}
```
Comment (lavalamp) [inline examples/workqueue/main.go]: select {} is a more elegant way to block forever

File: examples/workqueue/main.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		indexer:  indexer,
+		queue:    queue,
+		f:        f,
+	}
+}
+
+type ControllerFunc func(cache.Indexer, workqueue.RateLimitingInterface) bool
+
+func (c *Controller) Run(threadiness int, stopCh chan struct{}) {
+	defer NewPanicCatcher()
```
Comment (lavalamp) [inline examples/workqueue/main.go]: This line isn't doing what you think it's doing :)

File: examples/workqueue/main.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	// Create the connection
+	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
+	if err != nil {
+		panic(err.Error())
```
Comment (lavalamp) [inline examples/workqueue/main.go]: glog.Fatalf

File: examples/workqueue/main.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/golang/glog"
```
Comment (lavalamp) [inline examples/workqueue/main.go]: add blank lines so that there's three sections: builtin packages, k8s.io packages, 3rd party dependencies

**Comment by lavalamp** [comment]
thanks for the example!

**Comment by rmohr** [comment]
Thx for the review. Should have addressed all issues, but I am still not sure what the best practice is when we can't retrieve an item from the store in the process loop.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[examples/workqueue/main.go, line 22 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-22:-KaM8Z_3IwvyqxfzYn-7:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L22)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

add blank lines so that there's three sections: builtin packages, k8s.io packages, 3rd party dependencies
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-64:-KaM8gGanS5hRDxcLEVz:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L64)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

This line isn't doing what you think it's doing :)
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 89 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-89:-KaM8h7CYpoiQRR9JLbh:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L89)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

glog.Fatalf
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 138 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-138:-KaM8i92IY45vBRTb_ky:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L138)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

s/save/safe/
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 156 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-156:-KaM8jgzycK3yfPem5co:b8ztmyv) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L156)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

this line seems wrong
</blockquote></details>

Should be more correct now. Still don't know if this is a good practice what I do here ...

---

*[examples/workqueue/main.go, line 162 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-162:-KaM8pmygPEzqHBY9e7_:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L162)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

Use glog for log messages (or the built-in log package).
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 188 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-188:-KaM8vKex7Iur5AS49Sn:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L188)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

select {} is a more elegant way to block forever
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-:-KaM8zH-isjxp50dqrnR:brb7dok)*

File: examples/workqueue/main.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+var (
+	kubeconfig string
+	master     string
+)
+
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.StringVar(&master, "master", "", "master url")
+}
```
Comment (ash2k) [inline examples/workqueue/main.go]: Why not put flags and global variables into `main()`? Global state hurts code reusability/etc.

File: examples/workqueue/main.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+		queue:    queue,
+		f:        f,
+	}
+}
+
+type ControllerFunc func(cache.Indexer, workqueue.RateLimitingInterface) bool
+
+func (c *Controller) Run(threadiness int, stopCh chan struct{}) {
+	// Don't crash the loop on a panic
+	defer catchPanic()
```
Comment (ash2k) [inline examples/workqueue/main.go]: IMHO this is not relevant for this example. Also, again IMHO, it is not a great design choice :) Panic means something is completely wrong and there is no way to handle it. Why catch it?

**Comment by rmohr** [comment]
I signed it

**Comment by rmohr** [comment]
@k8s-ci-robot I signed it

**Comment by rmohr** [comment]
@k8s-ci-robot  I signed it

**Comment by rmohr** [comment]
@caesarxuchao @lavalamp anything more you want me to change?

**Comment by caesarxuchao** [comment]
@mbohlool it would be great if you can also take a look at the pull.

**Comment by mbohlool** [comment]
@rmohr @caesarxuchao Should this PR sent against main repo vs client-go repo?

**Comment by caesarxuchao** [comment]
The main repo please. It will be copied over to client-go.

**Comment by rmohr** [comment]
Ok, will do a PR against the main repo :+1:

**Comment by mbohlool** [comment]
@rmohr Thanks. Let me know when you send that PR, and I will review it with priority.

**Comment by ymruan** [comment]
workqueue is not in client-go 2.0, which kind of queue should I use?

**Comment by caesarxuchao** [comment]
@ymruan is it possible to use the workqueue in k8s.io/kubernetes release-1.5? I guess it will work.

**Pull request closed without merging**